File: frontend/src/api/serverOps.ts
import api from './client';
import { getAuth } from '../auth';

export type NodeLogKind = 'panel' | 'xray';
export type ServerHistoryMetric = 'cpu' | 'mem' | 'netUp' | 'netDown' | 'online' | 'load1' | 'load5' | 'load15';
export type ServerHistoryBucket = 2 | 30 | 60 | 180 | 360 | 720 | 1440 | 2880 | 10080;

export interface X25519KeyPair {
  privateKey?: string;
  publicKey?: string;
}

export interface Mldsa65KeyPair {
  privateKey?: string;
  publicKey?: string;
  [key: string]: unknown;
}

export interface VlessEncryptionAuth {
  id?: string | number;
  label?: string;
  encryption?: string;
  decryption?: string;
  [key: string]: unknown;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

function assertGeneratorPayload<T>(payload: T): T {
  if (isRecord(payload) && payload.error) {
    throw new Error(String(payload.error));
  }
  return payload;
}

export async function restartXray(nodeId: number): Promise<any> {
  const res = await api.post(`/v1/servers/${nodeId}/restart-xray`, {}, { auth: getAuth() });
  return res.data;
}

export async function generateNodeUuid(nodeId: number): Promise<string> {
  const res = await api.get(`/v1/nodes/${nodeId}/generate-uuid`, { auth: getAuth() });
  const payload = assertGeneratorPayload(res.data || {});
  const uuid = String(payload.uuid || '').trim();
  if (!uuid) {
    throw new Error('UUID generator returned an empty value');
  }
  return uuid;
}

export async function generateNodeX25519(nodeId: number): Promise<X25519KeyPair> {
  const res = await api.get(`/v1/nodes/${nodeId}/generate-x25519`, { auth: getAuth() });
  const payload = assertGeneratorPayload<X25519KeyPair & { error?: unknown }>(res.data || {});
  if (!payload.privateKey && !payload.publicKey) {
    throw new Error('X25519 generator returned an empty keypair');
  }
  return payload;
}

export async function generateNodeMldsa65(nodeId: number): Promise<Mldsa65KeyPair> {
  const res = await api.get(`/v1/nodes/${nodeId}/generate-mldsa65`, { auth: getAuth() });
  const payload = assertGeneratorPayload<Mldsa65KeyPair & { error?: unknown }>(res.data || {});
  if (!payload.privateKey && !payload.publicKey) {
    throw new Error('ML-DSA-65 generator returned an empty keypair');
  }
  return payload;
}

export async function generateNodeVlessEncryption(nodeId: number): Promise<VlessEncryptionAuth[]> {
  const res = await api.get(`/v1/nodes/${nodeId}/generate-vless-enc`, { auth: getAuth() });
  const payload = assertGeneratorPayload<unknown>(res.data || {});
  const rawAuths = Array.isArray(payload)
    ? payload
    : isRecord(payload) && Array.isArray(payload.auths)
      ? payload.auths
      : isRecord(payload) && Array.isArray(payload.obj)
        ? payload.obj
        : isRecord(payload)
          ? [payload]
          : [];
  const auths = rawAuths
    .map((item): VlessEncryptionAuth | null => {
      if (!isRecord(item)) {
        const value = String(item || '').trim();
        return value ? { label: value, encryption: value, decryption: value } : null;
      }
      return {
        ...item,
        id: item.id as string | number | undefined,
        label: String(item.label || item.id || '').trim(),
        encryption: String(item.encryption || '').trim(),
        decryption: String(item.decryption || '').trim(),
      };
    })
    .filter((item): item is VlessEncryptionAuth => item !== null);
  if (!auths.length) {
    throw new Error('VLESS encryption generator returned an empty payload');
  }
  return auths;
}

export async function getNodeLogs(
  nodeId: number,
  kind: NodeLogKind,
  options: { count?: number; level?: string } = {},
): Promise<string[]> {
  const endpoint = kind === 'xray'
    ? `/v1/nodes/${nodeId}/xray-logs`
    : `/v1/nodes/${nodeId}/server-logs`;
  const res = await api.get(endpoint, {
    params: { count: options.count ?? 200, level: options.level },
    auth: getAuth(),
  });
  const payload = res.data || {};
  if (payload.error) throw new Error(String(payload.error));
  return Array.isArray(payload.logs) ? payload.logs : [];
}

export async function getXrayVersions(nodeId: number): Promise<string[]> {
  const res = await api.get(`/v1/nodes/${nodeId}/xray-versions`, { auth: getAuth() });
  return res.data?.versions || [];
}

export async function installXray(nodeId: number, version: string): Promise<any> {
  const res = await api.post(`/v1/nodes/${nodeId}/install-xray/${version}`, {}, { auth: getAuth() });
  return res.data;
}

export async function stopXray(nodeId: number): Promise<any> {
  const res = await api.post(`/v1/nodes/${nodeId}/stop-xray`, {}, { auth: getAuth() });
  return res.data;
}

export async function updatePanel(nodeId: number): Promise<any> {
  const res = await api.post(`/v1/nodes/${nodeId}/update-panel`, {}, { auth: getAuth() });
  return res.data;
}

export async function updateGeofile(nodeId: number): Promise<any> {
  const res = await api.post(`/v1/nodes/${nodeId}/update-geofile`, {}, { auth: getAuth() });
  return res.data;
}

export async function getOutboundsTraffic(nodeId: number): Promise<any[]> {
  const res = await api.get(`/v1/nodes/${nodeId}/outbounds-traffic`, { auth: getAuth() });
  return res.data?.outbounds || [];
}

export async function getNodeTraffic(nodeId: number): Promise<any> {
  const res = await api.get(`/v1/nodes/${nodeId}/traffic`, { auth: getAuth() });
  return res.data;
}

export async function getNodeOnlineClients(nodeId: number): Promise<any> {
  const res = await api.get(`/v1/nodes/${nodeId}/online-clients`, { auth: getAuth() });
  return res.data;
}

export async function getXrayMetrics(nodeId: number): Promise<any> {
  const res = await api.get(`/v1/nodes/${nodeId}/xray-metrics`, { auth: getAuth() });
  return res.data;
}

export async function getXrayObservatory(nodeId: number): Promise<any> {
  const res = await api.get(`/v1/nodes/${nodeId}/xray-observatory`, { auth: getAuth() });
  return res.data;
}

export async function getXrayConfig(nodeId: number): Promise<any> {
  const res = await api.get(`/v1/nodes/${nodeId}/xray-config`, { auth: getAuth() });
  return res.data;
}

export async function getServerHistory(
  nodeId: number,
  metric: ServerHistoryMetric,
  bucket: ServerHistoryBucket = 360,
): Promise<Array<{t: number; v: number}>> {
  const res = await api.get(`/v1/nodes/${nodeId}/server-history/${metric}`, {
    params: { bucket },
    auth: getAuth(),
  });
  return res.data?.data || [];
}

export async function resetAllNodeTraffics(nodeId: number): Promise<any> {
  const res = await api.post(`/v1/inbounds/${nodeId}/reset-all-traffics`, {}, { auth: getAuth() });
  return res.data;
}

export async function getApiTokens(nodeId: number): Promise<any[]> {
  const res = await api.get(`/v1/nodes/${nodeId}/api-tokens`, { auth: getAuth() });
  return res.data?.tokens || res.data || [];
}

export async function createApiToken(nodeId: number, name: string): Promise<any> {
  const res = await api.post(`/v1/nodes/${nodeId}/api-tokens`, { name }, { auth: getAuth() });
  return res.data;
}

export async function deleteApiToken(nodeId: number, tokenId: number): Promise<any> {
  const res = await api.delete(`/v1/nodes/${nodeId}/api-tokens/${tokenId}`, { auth: getAuth() });
  return res.data;
}

export async function setApiTokenEnabled(nodeId: number, tokenId: number, enabled: boolean): Promise<any> {
  const res = await api.post(`/v1/nodes/${nodeId}/api-tokens/${tokenId}/set-enabled`, { enabled }, { auth: getAuth() });
  return res.data;
}

File: backend/routers/server_ops.py
"""Server operations router — history metrics, logs, API tokens, observatory."""
from typing import Dict
from fastapi import APIRouter, HTTPException, Request
from fastapi.concurrency import run_in_threadpool


def build_server_ops_router(
    *,
    check_auth,
    xui_monitor,
    server_monitor,
    get_node_or_404,
):
    router = APIRouter()

    async def _run(method, *args, **kwargs):
        return await run_in_threadpool(method, *args, **kwargs)

    async def _node(node_id):
        return await run_in_threadpool(get_node_or_404, node_id)

    @router.get("/api/v1/nodes/{node_id}/server-history/{metric}")
    async def get_server_history(
        request: Request, node_id: int, metric: str, bucket: str = "5m"
    ):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        node = await _node(node_id)
        return await _run(server_monitor.get_server_history, node, metric, bucket)

    @router.get("/api/v1/nodes/{node_id}/panel-update-info")
    async def get_panel_update_info(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        node = await _node(node_id)
        return await _run(server_monitor.get_panel_update_info, node)

    @router.get("/api/v1/nodes/{node_id}/xray-observatory")
    async def get_xray_observatory(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        node = await _node(node_id)
        return await _run(server_monitor.get_xray_observatory, node)

    @router.get("/api/v1/nodes/{node_id}/api-tokens")
    async def list_api_tokens(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        node = await _node(node_id)
        return await _run(server_monitor.get_api_tokens, node)

    @router.post("/api/v1/nodes/{node_id}/api-tokens")
    async def create_api_token(request: Request, node_id: int, data: Dict):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        name = (data.get("name") or "").strip()
        if not name:
            raise HTTPException(status_code=400, detail="name required")
        node = await _node(node_id)
        result = await _run(server_monitor.create_api_token, node, name)
        if "error" in result:
            raise HTTPException(status_code=500, detail=result["error"])
        return result

    @router.get("/api/v1/nodes/{node_id}/server-logs")
    async def get_server_logs(
        request: Request, node_id: int,
        count: int = 100, level: str = "info"
    ):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        node = await _node(node_id)
        return await _run(server_monitor.get_server_logs, node, count=min(max(count, 1), 500), level=level)

    @router.get("/api/v1/nodes/{node_id}/xray-logs")
    async def get_xray_logs(request: Request, node_id: int, count: int = 100, level: str = "info"):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        node = await _node(node_id)
        return await _run(server_monitor.get_xray_logs, node, count=min(max(count, 1), 500), level=level)

    @router.get("/api/v1/nodes/{node_id}/xray-versions")
    async def get_xray_versions(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.get_xray_versions, await _node(node_id))

    @router.post("/api/v1/nodes/{node_id}/install-xray/{version}")
    async def install_xray(request: Request, node_id: int, version: str):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.install_xray, await _node(node_id), version)

    @router.post("/api/v1/nodes/{node_id}/update-geofile")
    async def update_geofile(request: Request, node_id: int, data: Dict = None):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        file_name = (data or {}).get("fileName", "")
        return await _run(server_monitor.update_geofile, await _node(node_id), file_name)

    @router.post("/api/v1/nodes/{node_id}/stop-xray")
    async def stop_xray(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        ok = await _run(server_monitor.stop_xray, await _node(node_id))
        return {"success": ok}

    @router.post("/api/v1/nodes/{node_id}/update-panel")
    async def update_panel(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.update_panel, await _node(node_id))

    @router.get("/api/v1/nodes/{node_id}/xray-metrics")
    async def get_xray_metrics(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.get_xray_metrics, await _node(node_id))

    @router.get("/api/v1/nodes/{node_id}/outbounds-traffic")
    async def get_outbounds_traffic(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.get_outbounds_traffic, await _node(node_id))

    @router.get("/api/v1/nodes/{node_id}/generate-uuid")
    async def generate_uuid(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.generate_uuid, await _node(node_id))

    @router.get("/api/v1/nodes/{node_id}/generate-x25519")
    async def generate_x25519(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.generate_x25519_cert, await _node(node_id))

    @router.get("/api/v1/nodes/{node_id}/generate-vless-enc")
    async def generate_vless_enc(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.generate_vless_enc, await _node(node_id))

    @router.get("/api/v1/nodes/{node_id}/generate-mldsa65")
    async def generate_mldsa65(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.generate_mldsa65, await _node(node_id))

    @router.get("/api/v1/nodes/{node_id}/xray-config")
    async def get_xray_config(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.get_xray_config, await _node(node_id))

    @router.post("/api/v1/nodes/{node_id}/backup-telegram")
    async def backup_telegram(request: Request, node_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        return await _run(server_monitor.backup_to_telegram, await _node(node_id))

    @router.delete("/api/v1/nodes/{node_id}/api-tokens/{token_id}")
    async def delete_api_token(request: Request, node_id: int, token_id: int):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        ok = await _run(server_monitor.delete_api_token, await _node(node_id), token_id)
        return {"success": ok}

    @router.post("/api/v1/nodes/{node_id}/api-tokens/{token_id}/set-enabled")
    async def set_api_token_enabled(request: Request, node_id: int, token_id: int, data: Dict):
        user = check_auth(request)
        if not user:
            raise HTTPException(status_code=401)
        ok = await _run(server_monitor.set_api_token_enabled,
            await _node(node_id),
            token_id,
            bool(data.get("enabled", True)),
        )
        return {"success": ok}

    return router
